### DISCUSSION TITLE
top-k only limited to 1024 in GPU_IVF_PQ index

### DISCUSSION THREAD

**Discussion by akcopy** [milvus-io/milvus#30333] [Q&A and General discussion] [answered]
Based on the docs, it appears that I can have a top-k of up to 8192 with a GPU_IVF_PQ index. However, recently it appears that I'm only limited to a topk of 1024.  Do you know how I can go back to a top-k of 8192? See error message below:

failed to search/query delegator 60 for channel by-dev-rootcoord-dml_0_447059044643680506v0: fail to Search, QueryNode ID=60, reason=worker(60) query failed: UnexpectedError: Assert "failed to search: out of range in json: param k out of range [ 1,1024 ]" at /go/src/github.com/milvus-io/milvus/internal/core/src/index/VectorMemIndex.cpp:326:

**Comment by presburger** [comment]
@akcopy Hi, You can try using GPU brute force search; GPU brute force search currently supports a size of up to 8096 for top-k. "GPU_BRUTE_FORCE"

**Reply by akcopy** [reply]
Isn't there only 2 types of GPU indexes, GPU_IVF_FLAT and GPU_IVF_PQ? IVF_FLAT is worse because the max topk is only 256...

**Reply by yhmo** [reply, marked as answer]
Topk of GPU_IVF_PQ has been limited to 1024 recently due to the limitation of a third-party lib. Currently not able to enlarge this value.
"GPU_BRUTE_FORCE" is a special "index type" in Milvus v2.4.x, topk is not limited.

**Reply by akcopy** [reply]
Thanks for the clarification. What are the properties of the GPU_BRUTE_FORCE index, is there a page explaining it? For eg index and search parameters, memory usage, speed compared to HNSW, etc...

**Reply by akcopy** [reply]
Is there a chance that I can use a slightly older version of Milvus that supports topk of 8192? It's important for my application. Which version was it changed in?

**Reply by yhmo** [reply]
v2.4 is not released yet, no document for GPU_BRUTE_FORCE. It is simply a brute-force search by GPU. Definitely, the search performance is slower than other GPU index.

The first version to support GPU index is v2.3.0, you can try this version.

**Reply by akcopy** [reply]
In that case it won't be useful to me - if it's slow I might as well just use HNSW or SCANN. The reason why I want GPU_IVF_PQ is because of the speed and low memory sizing requirements. Is there a chance that the top-k be reverted to 8192 in future versions? I will try v2.3 again to see. I remember that I did use top-k of 8192 a few months before - it stopped working after I updated the image.

**Reply by yhmo** [reply]
The latest version uses the NVIDIA RAFT lib for GPU search: https://developer.nvidia.com/blog/accelerating-vector-search-using-gpu-powered-indexes-with-rapids-raft/

**Reply by akcopy** [reply]
I tried to revert back to v2.3.0 (deleting my docker containers and then reloading using docker compose with v2.3.0-gpu set in my .yaml file) but no matter what I try, top-k appears to be locked to 1024 now. 

I'm getting this error.
pymilvus.exceptions.MilvusException: <MilvusException: (code=0, message=attempt #0: fail to Search, QueryNode ID=62, reason=worker(62) query failed: UnexpectedError: Assert "failed to search: out of range in json: param k out of range [ 1,1024 ]" at /go/src/github.com/milvus-io/milvus/internal/core/src/index/VectorMemIndex.cpp:319: channel=by-dev-rootcoord-dml_0_447059044643680506v0: fail to search/query shard leader: attempt #1: no available shard delegator found: service unavailable: fail to search on all shard leaders)> 

Would you be able to help on this?

**Comment by xiaofan-luan** [comment]
this is nothing to do with milvus and there is no way to relax the limitation.
With 2.4 your can use GPU brute force
